src/network/rpc/base.cpp

```diff
@@ -66,30 +78,93 @@ void rpc_server::add(const string &name, const pfi::lang::shared_ptr<invoker_bas
 
 bool rpc_server::serv(uint16_t port, int nthreads)
 {
-  pfi::lang::shared_ptr<server_socket> ssock(new server_socket());
-  if (!ssock->create(port))
+  if (!start(port, nthreads)) {
+    return false;
+  }
+  wait_until_stopped();
+  return true;
+}
+
+bool rpc_server::start(uint16_t port, int nthreads)
+{
+  socket = pfi::lang::shared_ptr<server_socket>(new server_socket());
+  if (!socket->create(port))
     return false;
 
-  port_num = ssock->port();
+  port_num = socket->port();
 
-  vector<pfi::lang::shared_ptr<thread> > ths(nthreads);
-  for (int i=0; i<nthreads; i++){
-      ths[i]=pfi::lang::shared_ptr<thread>(new thread(bind(&rpc_server::process, this, ssock)));
-    if (!ths[i]->start()) return false;
+  {
+    std::lock_guard<std::mutex> lock(state_mutex);
+    // NOTE: state should be set as `running` before starting threads
+    //       because each thread runs while the state is `running`.
+    state = server_state::RUNNING;
+    threads = vector<pfi::lang::shared_ptr<thread>>(nthreads);
+    for (int i=0; i<nthreads; i++){
+      threads[i]=pfi::lang::shared_ptr<thread>(new thread(bind(&rpc_server::process, this, socket)));
+      if (!threads[i]->start()) return false;
+      ++num_running_threads;
+    }
   }
-  for (int i=0; i<nthreads; i++)
-    ths[i]->join();
   return true;
 }
 
+void rpc_server::stop()
+{
+  std::lock_guard<std::mutex> lock(state_mutex);
```

Comment:
`stop()` acquire lock with `state_mutex` , and invoke some RPC requests.
TheseRPC requests wait the result of server thread's process.
And a server thread's process check `is_running()` at the begging.
`is_running()` try to acqurie lock with `state_mutex`, but this thread has already acquired.
So this lock_gurad may be the cause of the dead-lock.

One example solution for dead-lock, release this lock after change state.

Reply:
When no thread wait on accept, RPC request raise exception immediately. So no dead lock will occur.